core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -0,0 +1,1152 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.netty.handler.codec.http2.Http2Headers.PseudoHeaderName.hasPseudoHeaderFormat;
+import static io.netty.util.internal.MathUtil.findNextPositivePowerOfTwo;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import io.netty.handler.codec.DateFormatter;
+import io.netty.util.AsciiString;
+
+/**
+ * The base container implementation of HTTP/2 headers.
+ */
+class HttpHeadersBase implements HttpHeaderGetters {
+
+    static final int DEFAULT_SIZE_HINT = 16;
+
+    /**
+     * Constant used to seed the hash code generation. Could be anything but this was borrowed from murmur3.
+     */
+    static final int HASH_CODE_SEED = 0xc2b2ae35;
+
+    // XXX(anuraaga): It could be an interesting future optimization if we can use something similar
+    //                to an EnumSet when it's applicable, with just one each of commonly known header names.
+    //                It should be very common.
+    @VisibleForTesting
+    final HeaderEntry[] entries;
```

Comment:
Just wondering, did you consider guava's `Multimap`? I think this implementation is similar to a `LinkedListMultimap`. It might still be better to hand-implement, just want to make sure you're aware of it :)

Reply:
I knew about `Multimap` but I didn't use it because:

- We have to do a case-insensitive search for any `CharSequence` keys (both `AsciiString` and `String`)
- We cannot expose Guava API in our public API because we shade Guava.
  - Or extra allocation cost if we wrap a `MultiMap`.